packages/gatsby/src/schema/node-model.js

```diff
@@ -278,6 +483,254 @@ const toNodeTypeNames = (schema, gqlTypeName) => {
     .map(type => type.name)
 }
 
+const getQueryFields = ({ filter, sort, group, distinct }) => {
+  const filterFields = filter ? dropQueryOperators(filter) : {}
+  const sortFields = (sort && sort.fields) || []
+
+  if (group && !Array.isArray(group)) {
+    group = [group]
+  } else if (group == null) {
+    group = []
+  }
+
+  if (distinct && !Array.isArray(distinct)) {
+    distinct = [distinct]
+  } else if (distinct == null) {
+    distinct = []
+  }
+
+  return merge(
+    filterFields,
+    ...sortFields.map(pathToObject),
+    ...group.map(pathToObject),
+    ...distinct.map(pathToObject)
+  )
+}
+
+const pathToObject = path => {
+  if (path && typeof path === `string`) {
+    return path.split(`.`).reduceRight((acc, key) => {
+      return { [key]: acc }
+    }, true)
+  }
+  return {}
+}
+
+const dropQueryOperators = filter =>
+  Object.keys(filter).reduce((acc, key) => {
+    const value = filter[key]
+    const k = Object.keys(value)[0]
+    const v = value[k]
+    if (_.isPlainObject(value) && _.isPlainObject(v)) {
+      acc[key] =
+        k === `elemMatch` ? dropQueryOperators(v) : dropQueryOperators(value)
+    } else {
+      acc[key] = true
+    }
+    return acc
+  }, {})
+
+const mergeObjects = (obj1, obj2) =>
```

Comment:
i wonder if we can leverage lodash helpers instead.

Reply:
Seems we can.